streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreRestoreWriteBuffer.java

```diff
@@ -0,0 +1,269 @@
+                            new KeyValue<>(segmentEntry.getKey().get(), segmentEntry.getValue()),
+                            segmentsBatch);
+                    }
+                }
+
+                // write to db. all the logical segments share the same physical store,
+                // so we can use any segment to perform the write
+                allSegments.get(0).dbSegment().write(segmentsBatch);
+            }
+        } catch (final RocksDBException e) {
+            log.error("Error restoring batch to RocksDBVersionedStore segments store.");
+            throw e;
+        }
+        segmentsWriteBuffer.clear();
+
+        // flush latest value store
+        try (final WriteBatch latestValueBatch = new WriteBatch()) {
+            // collect entries into write batch
+            for (final Map.Entry<Bytes, Optional<byte[]>> latestValueEntry : latestValueWriteBuffer.entrySet()) {
+                final byte[] value = latestValueEntry.getValue().orElse(null);
+                dbClient.addToLatestValueBatch(
+                    new KeyValue<>(latestValueEntry.getKey().get(), value),
+                    latestValueBatch);
+            }
+
+            // write to db
+            dbClient.writeLatestValues(latestValueBatch);
+        } catch (final RocksDBException e) {
+            log.error("Error restoring batch to RocksDBVersionedStore latest value store.");
+            throw e;
+        }
+        latestValueWriteBuffer.clear();
+    }
+
+    /**
+     * The object representation of the write buffer corresponding to a single segment store.
+     * Contains the write buffer itself (a simple hash map) and also a reference to the underlying
+     * persistent segment store.
+     */
+    private class WriteBufferSegmentWithDbFallback implements VersionedStoreSegment {
+
+        private final long id;
+        private final Map<Bytes, byte[]> data;
+        private final LogicalKeyValueSegment dbSegment;
+
+        WriteBufferSegmentWithDbFallback(final LogicalKeyValueSegment dbSegment) {
+            this.dbSegment = Objects.requireNonNull(dbSegment);
+            this.id = dbSegment.id();
+            this.data = new HashMap<>();
+
+            // register segment with segments store
+            segmentsWriteBuffer.put(id, this);
+        }
+
+        LogicalKeyValueSegment dbSegment() {
+            return dbSegment;
+        }
+
+        @Override
+        public long id() {
+            return id;
+        }
+
+        @Override
+        public void put(final Bytes key, final byte[] value) {
+            // all writes go to the write buffer
+            data.put(key, value);
+        }
+
+        @Override
+        public byte[] get(final Bytes key) {
+            final byte[] bufferValue = data.get(key);
+            if (bufferValue != null) {
+                return bufferValue;
+            }
+            return dbSegment.get(key);
+        }
+
+        Set<Entry<Bytes, byte[]>> getAll() {
```

Comment:
Why return `Set<Entry>` here and not just the `data` `Map`?

Reply:
Minor optimization to not need to copy the map. (If we were to return a map, I don't think the return value should be modifiable.) If you prefer copying the map for readability we can do that too.

Reply:
I don't think that a deep copy is required to guard the map from modification (`return Collections.unmodifiableMap();` would do the trick). Given the limited scope who calls `getAll()` it should be fine to just put the guard in the impl without making it a contract of the API (we do this often inside KS code base).

Reply:
OK, I'll make the update.